### ISSUE TITLE
XGBoostClassificationModel with continuous label may fail to encode

### ISSUE THREAD

**Issue by shiyuan** [jpmml/jpmml-sparkml-xgboost#8]
Hi,

I'm trying to convert a tree-label XGBoostClassificationModel to PMML.
```
jpmml-sparkml-xgboost: 0feb4d4e45e94cae868e108015f85bf64d2dea0a
jpmml-sparkml: 1.2.12
jpmml-xgboost: 1.3.2
jpmml-converter: 1.3.2
xgboost4j-spark: 0.7
```

The string label works well, but the double label meets the exception that the label.size < 3 throws an illegalArgumentException in [jpmml.converter.mining.MiningModelUtils](https://github.com/jpmml/jpmml-converter/blob/1.3.2/src/main/java/org/jpmml/converter/mining/MiningModelUtil.java#L75)   

Then I find the bug in [jpmml.sparkml.ModelConverter](https://github.com/jpmml/jpmml-sparkml/blob/1.2.12/src/main/java/org/jpmml/sparkml/ModelConverter.java#L98), that for continuous label, the XGBoostClassificationModel can not detect the numClasses, I add the code below, and it works well, avoid the exception above.
```
if (model instanceof XGBoostClassificationModel) {
    XGBoostClassificationModel xgbModel = (XGBoostClassificationModel) model;
    numClasses = xgbModel.numClasses();
}
```
Hope it helps.

**Comment by vruusmann** [comment]
> The string label works well, but the double label
> meets the exception that the label.size < 3

The best way to "fix" this issue would be to update your Apache Spark pipeline so that the column representing the target field would be detected as `CategoricalFeature` (not `ContinuousFeature`). By default, the feature type is derived from data type - strings are assumed to be discrete/categorical, while doubles are assumed to be continuous.

So, to make things work out-of-the-box, try casting the target column to string (target category values could be simply "0", "1" and "2" for a three-class classification problem), and then apply `StringIndexer` to this column.

> the XGBoostClassificationModel can not detect the numClasses

If I remember correctly, then the problem is that `XGBoostClassificationModel` does not inherit from Apache Spark's `ClassificationModel`, so this check `(model instanceof ClassificationModel)` does not fire correctly.

I don't want to introduce the `(model instanceof XGBoostClassificationModel)` check into the JPMML-SparkML library codebase, because this class is coming from a third-party library (xgboost4j), not from Apache Spark core libraries.

**Comment by vruusmann** [comment]
> The string label works well, but the double label
> meets the exception that the label.size < 3

Alternatively, this issue could be "fixed" if there were "domain definition" meta-transformation classes available as detailed in this feature request: https://github.com/jpmml/jpmml-sparkml/issues/14

There would be `CategoricalDomain` and `ContinuousDomain` classes available (just like in the SkLearn2PMML package), and one could re-classify a double column from continuous optype to categorical optype simply by applying the `CategoricalDomain` meta-transformation to it.

**Comment by shiyuan** [comment]
Get it, I fixed my code using the casting way.

But I think, the model not instanceof ClassificationModel([ModelConverter](https://github.com/jpmml/jpmml-sparkml/blob/1.2.12/src/main/java/org/jpmml/sparkml/ModelConverter.java#L98)) should throw an exception for unknown model, not just treated as 2-class classification model.

**Comment by vruusmann** [comment]
> the model not instanceof ClassificationModel should throw an exception for unknown model

That's a valid point. However, maybe the JPMML-SparkML library should attempt a few other "hacks" before throwing the exception.

For example, perhaps it should use Java reflection to check if the model object has a `numClasses()` field/method.

Of course, the best thing to happen would be if the XGBoost4J library corrected its class hierarchy so that `XGBoostClassificationModel` would be a subclass of `org.apache.spark.ml.classification.ClassificationModel` (actually `ProbabilisticClassificationModel`). Any volunteers to open this issue with the XGBoost project?

**Comment by shiyuan** [comment]
> Any volunteers to open this issue with the XGBoost project

It seems there is no need to do so.

In `xgboost4j-spark-0.70`, `XGBoostClassificationModel` does not inherit from `ClassificationModel`

But in `xgboost4j-spark-0.80`,  `XGBoostClassificationModel`  [extends](https://github.com/dmlc/xgboost/blob/release_0.80/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala#L212) from `ProbabilisticClassificationModel`, then [extends](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala#L80) from `ClassificationModel`, it's already satisfied.

Thanks.
